nntrainer/tensor/tensor.cpp

```diff
@@ -2148,1680 +734,401 @@ Tensor &Tensor::dot_batched_deriv_wrt_2(Tensor &m_deriv,
+  itensor->zoneout_mask(opposite, zoneout);
+}
 
-    result.resize(batch_size);
+std::vector<Tensor> Tensor::split(unsigned num_size, int axis) {
+  NNTR_THROW_IF(num_size == 0, std::invalid_argument)
+    << "num size cannot be zero";
 
-    for (unsigned int b = 0; b < batch_size; b++) {
-      auto max_iter =
-        std::max_element(data + b * feature_len, data + (b + 1) * feature_len);
-      result[b] = std::distance(data, max_iter) - (b * feature_len);
-    }
+  if (axis == -1) {
+    axis = 3;
   }
-  if (getDataType() == ml::train::TensorDim::DataType::FP16) {
-#ifdef ENABLE_FP16
-    const _FP16 *data = getData<_FP16>();
-    size_t batch_size = batch();
-    size_t feature_len = dim.getFeatureLen();
 
-    result.resize(batch_size);
+  NNTR_THROW_IF(!(0 <= axis && axis < 4), std::invalid_argument)
+    << "cannot split axis of axis: " << axis;
 
-    for (unsigned int b = 0; b < batch_size; b++) {
-      auto max_iter =
-        std::max_element(data + b * feature_len, data + (b + 1) * feature_len);
-      result[b] = std::distance(data, max_iter) - (b * feature_len);
-    }
-#else
-    throw std::invalid_argument("Error: enable-fp16 is not enabled");
-#endif
-  }
+  NNTR_THROW_IF(getDim().getTensorDim(axis) % num_size != 0,
+                std::invalid_argument)
+    << "axis is not divisible by num_size, axis: " << axis
+    << " num size: " << num_size;
 
-  return result;
-}
+  std::vector<size_t> sizes;
+  sizes.resize(num_size);
 
-int Tensor::erf_i() {
-  erf(*this);
-  return ML_ERROR_NONE;
+  unsigned int sz = getDim().getTensorDim(axis) / num_size;
+  std::fill(sizes.begin(), sizes.end(), sz);
+
+  return split(sizes, axis);
 }
 
-Tensor Tensor::erf() const {
-  Tensor t;
-  return erf(t);
+std::vector<Tensor> Tensor::split(std::vector<size_t> sizes, int axis) {
+  NNTR_THROW_IF(sizes.size() == 0, std::invalid_argument)
+    << "num size cannot be zero";
+
+  NNTR_THROW_IF(!(-1 <= axis && axis < 4), std::invalid_argument)
+    << "cannot split axis of axis: " << axis;
+
+  NNTR_THROW_IF(
+    std::any_of(sizes.begin(), sizes.end(), [](size_t sz) { return !sz; }),
+    std::invalid_argument)
+    << "among given sizes at least one of size is 0";
+
+  return itensor->split(sizes, axis);
 }
 
-Tensor &Tensor::erf(Tensor &out) const {
-  if (dim.getDataType() == ml::train::TensorDim::DataType::FP32) {
-    auto f = [](float in) { return std::erf(in); };
-    apply<float>(f, out);
-  } else if (dim.getDataType() == ml::train::TensorDim::DataType::FP16) {
+Tensor Tensor::cat(const std::vector<Tensor> &tensors, int axis) {
+  NNTR_THROW_IF(!(-1 <= axis && axis < 4), std::invalid_argument)
+    << "cannot split axis of axis: " << axis;
+
+  NNTR_THROW_IF(tensors.empty(), std::invalid_argument)
+    << "given tensor vector is empty";
+
+  Tensor output;
+  Tdatatype dtype = tensors.front().getDim().getDataType();
+
+  if (dtype == Tdatatype::FP32) {
+    output = FloatTensor::cat(tensors, axis);
```

Comment:
Do we need this if branch?

Reply:
I believe so. As far as I know, this function concatenates a vector of tensors by a given axis. 
However, the problem with this function is tensors are concatenated to the first tensor of the list, not the tensor that calls this function. Therefore, the FP16 tensor can concatenate a list of FP32 tensors.
While the input tensors have no relation with the tensor that calls this function, we need a decision point on whether to concatenate a list of FP32 tensors or FP16 tensors, and this should be made in the tensor class. 

I think there is plenty of room to improve the function itself and we can later newly design this function.

Reply:
But we change the api.. like tensors[0]->itensor.cat(tensors, axis)?

Reply:
we cannot use it as `tensors[0]->itensor.cat(tensors, axis)` since `itensor` is a private variable. I'll modify this in another way and let you know.

Reply:
I fixed this by creating a new member function of Tensor `concat()`, and the current `Tensor::cat()`, which is a static function, uses `concat()`. This should work exactly the same as how `tensors[0]->itensor.cat(tensors, axis)` would operate.